analyzer/block/block.go

```diff
@@ -0,0 +1,349 @@
+				ctx,
+				queries.QueueProcessingBlockRange,
+				height,
+				futureHeight,
+				b.analyzerName,
+			)
+			if err != nil {
+				backoff.Failure()
+				b.logger.Error("error queueing block range", "err", err)
+				continue
+			}
+			rows.Close()
+
+			height = futureHeight
+			// Timeout a bit to avoid spamming the database to much.
+			time.Sleep(6 * time.Second)
+			continue
+		}
+
+		if err := b.processor.HasBlock(ctx, height); err != nil {
+			backoff.Failure()
+			b.logger.Error("checking for block", "height", height, "err", err)
+			continue
+		}
+
+		// Insert into processing queue.
+		rows, err := b.target.Query(
+			ctx,
+			queries.QueueProcessingBlock,
+			height,
+			b.analyzerName,
+		)
+		if err != nil {
+			backoff.Failure()
+			b.logger.Error("error queueing block", "err", err)
+			continue
+		}
+		rows.Close()
+
+		backoff.Success()
+		height++
+	}
+}
+
+// unlockBlock unlocks a block.
+func (b *blockBasedAnalyzer) unlockBlock(ctx context.Context, height uint64) {
+	rows, err := b.target.Query(
+		ctx,
+		queries.UnlockBlockForProcessing,
+		b.analyzerName,
+		height,
+	)
+	if err != nil {
+		rows.Close()
+	}
+}
+
+// fetchBatchForProcessing fetches a batch of blocks for processing.
+func (b *blockBasedAnalyzer) fetchBatchForProcessing(ctx context.Context, from, to uint64) ([]uint64, error) {
+	var (
+		heights []uint64
+		rows    pgx.Rows
+		err     error
+	)
+
+	switch to {
+	case 0:
+		rows, err = b.target.Query(
+			ctx,
+			queries.PickBlocksForProcessing,
+			b.analyzerName,
+			from,
+			lockExpiryMinutes,
+			blocksBatchSize,
+		)
+	default:
+		rows, err = b.target.Query(
+			ctx,
+			queries.PickBlocksForProcessingRange,
+			b.analyzerName,
+			from,
+			to,
+			lockExpiryMinutes,
+			blocksBatchSize,
+		)
+	}
+	switch err {
+	case nil:
+		// Continues below.
+	case pgx.ErrNoRows:
+		// No blocks to process.
+		return heights, nil
+	default:
+		return nil, err
+	}
+
+	defer rows.Close()
+	for rows.Next() {
+		var height uint64
+		if err = rows.Scan(
+			&height,
+		); err != nil {
+			return nil, fmt.Errorf("scanning returned height: %w", err)
+		}
+		heights = append(heights, height)
+	}
+	return heights, nil
+}
+
+// Start starts the block analyzer.
+func (b *blockBasedAnalyzer) Start() {
+	ctx := context.Background()
+
+	// Run prework.
+	if err := b.processor.PreWork(ctx); err != nil {
+		b.logger.Error("prework failed", "err", err)
+		return
+	}
+
+	// Configure start height.
+	var startHeight uint64
+	latest, err := b.latestProcessedBlock(ctx)
+	switch err {
+	case nil:
+		b.logger.Debug("setting height using latest block", "height", latest+1)
+		startHeight = latest + 1
+	case pgx.ErrNoRows:
+		b.logger.Debug("setting height using range config", "height", b.config.From)
+		startHeight = b.config.From
+	default:
+		b.logger.Error("last block height not found",
+			"err", err,
+		)
+		return
+	}
+
+	// Start a worker to follow and enqueue incoming blocks.
+	go b.followIncoming(ctx, startHeight, b.config.To)
```

Comment:
Implemented this mega-query now. Also includes the configured block range.

Reply:
One more thing to add:
I think the analyzer should also query "latest available height" on the source and not lock/pick-up blocks above it. That way, it won't lock blocks it cannot process at the moment.

I will replace `SourceHasBlock`  with a method to fetch latest block as you already suggested in one comment. 

---

This would mostly be problematic if there would be multiple analyzer instances and one of the nodes would be behind others.

Reply:
This is now implemented